packages/astro/src/prefetch/index.ts

```diff
@@ -216,7 +218,14 @@ export function prefetch(url: string, opts?: PrefetchOptions) {
 	const priority = opts?.with ?? 'link';
 	debug?.(`[astro] Prefetching ${url} with ${priority}`);
 
-	if (priority === 'link') {
+	if (
+		clientPrerender &&
+		HTMLScriptElement.supports &&
+		HTMLScriptElement.supports('speculationrules')
+	) {
+		// this code is tree-shaken if unused
+		appendSpeculationRules(url);
```

Comment:
Maybe in the future we can use prerender/prefetch when using `with: 'link'` or `with: 'fetch'` respectively. That way links that enter the viewport don't get prerendered entirely. But we can improve this later if needed.